### DISCUSSION TITLE
Running cmds from javascript.

### DISCUSSION THREAD

**Discussion by pietru2004** [pocketbase/pocketbase#5620] [Q&A] [answered]
Is it possible to run cmds such as zip as java? Asking cause I want to make system that will start jar server. Idea is it would check in pb instance has newer version of jar, if it has download it else run currently downloaded version. Also I want to run zip via cron(Ik how to add cron to pb hooks, but not how to run cmd.)

**Comment by ganigeorgiev** [comment, marked as answer]
I don't understand what is the command you wan to run or how it is expected to execute but you can try the [`$os.cmd()`](https://pocketbase.io/jsvm/functions/_os.cmd.html) helper (make sure that any untrusted user arguments are properly sanitized/normalized to avoid remote code execution (RCE)).

**Reply by pietru2004** [reply]
I just want to run zip mkdir and java cmds

**Reply by dschissler** [reply]
>  Idea is it would check in pb instance has newer version of jar,

This leads me to believe that you're very confused about something.  Java Jar is a system command that would be installed at the OS system level.  PB provides a way to "shell out" to the system to run commands.

Typically these shell commands are to run a script or command to batch process something, and then to return.  I don't know if it's a good idea to have a long running service running as a shell command.

Maybe there is a Golang library that provides better integration with the shell or with Java directly that you could then provide to the PB Gojo VM, but you'd have to do that in Golang code.  It would then be available in the JS hooks.  I'm doing this for some specialized tech that I want included in my JS hooks that isn't relevant to PB as a whole.

Another idea would be that if you're using Linux then you can just install the Java server using `systemd` service files.  It's fairly easy to do, especially if there are already examples for that server tech.  It's basically the way to run Linux server procs now and so if the project is popular then it shouldn't be difficult to find.

(cc @ganigeorgiev) I've found though that using PB as a way to do system provisioning isn't great since PB really wants to create a data directory, and so it's not suitable for embedding system provisioning (Cobra) commands that will do this as root or whatever other user.  It would be nice if bootstrapping was a bit more flexible in this way.

In short, you're probably trying to combine too many tools and servers together.

**Reply by pietru2004** [reply]
@dschissler I am trying smthg like that

```
        console.log("Launching server...")
        var launch_jar = $os.cmd("java","-Duser.dir=$(pwd)/game_server","-jar","./game_server/server.jar",">>","log.txt")
        launch_jar.start()
        launch_jar.wait()
```

**Reply by pietru2004** [reply]
idea is to download latest jar and then run it... I can use docker compose or secondary sh script to run correct jar while I use pocketbase to download file

**Reply by dschissler** [reply]
It's ok, but if you plan on it being more serious then you can invest into doing it more formally.

**Reply by pietru2004** [reply]
Atm I want smthg that will just work
